File: HubPrediction/ChicagoTime.swift
import Foundation

enum ChicagoTime {
    static let tz = TimeZone(identifier: "America/Chicago") ?? .current

    private static func parts(_ ms: Double) -> DateComponents {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = tz
        return cal.dateComponents(
            [.year, .month, .day, .hour, .minute, .second, .weekday],
            from: Date(timeIntervalSince1970: ms / 1000.0)
        )
    }

    static func chicagoOffsetMs(_ ms: Double) -> Double {
        let p = parts(ms)
        let asUtc = Date.utcMs(y: p.year ?? 0, m: p.month ?? 1, d: p.day ?? 1, h: p.hour ?? 0, min: p.minute ?? 0, s: p.second ?? 0)
        return asUtc - ms
    }

    static func dayKey(_ ms: Double) -> String {
        let p = parts(ms)
        return String(format: "%04d-%02d-%02d", p.year ?? 0, p.month ?? 0, p.day ?? 0)
    }

    static func weekdayName(_ ms: Double) -> String {
        let f = DateFormatter()
        f.timeZone = tz
        f.dateFormat = "EEE"
        return f.string(from: Date(timeIntervalSince1970: ms / 1000.0))
    }

    static func formatClock(_ ms: Double) -> String {
        let f = DateFormatter()
        f.timeZone = tz
        f.dateFormat = "h:mm a"
        return f.string(from: Date(timeIntervalSince1970: ms / 1000.0))
    }

    static func formatDayLabel(_ ms: Double) -> String {
        let f = DateFormatter()
        f.timeZone = tz
        f.dateFormat = "EEE MMM d"
        return f.string(from: Date(timeIntervalSince1970: ms / 1000.0))
    }

    static func startOfChicagoDay(_ ms: Double) -> Double {
        let p = parts(ms)
        let guess = Date.utcMs(y: p.year ?? 0, m: p.month ?? 1, d: p.day ?? 1, h: 6, min: 0, s: 0)
        let off = chicagoOffsetMs(guess)
        return Date.utcMs(y: p.year ?? 0, m: p.month ?? 1, d: p.day ?? 1, h: 0, min: 0, s: 0) - off
    }

    static func align15(_ ms: Double) -> Double {
        let p = parts(ms)
        let minute = ((p.minute ?? 0) / 15) * 15
        let guess = Date.utcMs(y: p.year ?? 0, m: p.month ?? 1, d: p.day ?? 1, h: p.hour ?? 0, min: minute, s: 0)
        let off = chicagoOffsetMs(guess)
        return Date.utcMs(y: p.year ?? 0, m: p.month ?? 1, d: p.day ?? 1, h: p.hour ?? 0, min: minute, s: 0) - off
    }

    static func addChicagoDays(_ ms: Double, days: Int) -> Double {
        startOfChicagoDay(ms + Double(days) * HubMs.day + 12.0 * HubMs.hour)
    }

    static func slotsForDay(_ dayMs: Double) -> [Double] {
        let start = startOfChicagoDay(dayMs)
        return (0..<96).map { start + Double($0) * 15.0 * HubMs.minute }
    }

    static func upcomingDays(now: Double = Date.nowMs, count: Int = 7) -> [(t: Double, key: String, label: String)] {
        let today = startOfChicagoDay(now)
        return (0..<count).map { i in
            let t = addChicagoDays(today, days: i)
            return (t, dayKey(t), formatDayLabel(t))
        }
    }
}

extension Date {
    static var nowMs: Double { Date().timeIntervalSince1970 * 1000.0 }

    static func utcMs(y: Int, m: Int, d: Int, h: Int, min: Int, s: Int) -> Double {
        var c = DateComponents()
        c.year = y
        c.month = m
        c.day = d
        c.hour = h
        c.minute = min
        c.second = s
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(secondsFromGMT: 0) ?? .gmt
        return (cal.date(from: c) ?? Date()).timeIntervalSince1970 * 1000.0
    }
}

File: hub-prediction/src/lib/chicago-time.ts
const TZ = 'America/Chicago'

function parts(ms: number) {
  const map: Record<string, string> = {}
  for (const p of new Intl.DateTimeFormat('en-US', {
    timeZone: TZ,
    weekday: 'short',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false,
  }).formatToParts(new Date(ms))) {
    if (p.type !== 'literal') map[p.type] = p.value
  }
  const hour = map.hour === '24' ? 0 : Number(map.hour)
  return {
    weekday: map.weekday ?? '',
    year: Number(map.year),
    month: Number(map.month),
    day: Number(map.day),
    hour,
    minute: Number(map.minute),
    second: Number(map.second),
  }
}

export function chicagoOffsetMs(ms: number) {
  const p = parts(ms)
  const asUtc = Date.UTC(p.year, p.month - 1, p.day, p.hour, p.minute, p.second)
  return asUtc - ms
}

export function dayKey(ms: number) {
  const p = parts(ms)
  const m = String(p.month).padStart(2, '0')
  const d = String(p.day).padStart(2, '0')
  return `${p.year}-${m}-${d}`
}

export function weekdayName(ms: number) {
  return parts(ms).weekday
}

export function formatClock(ms: number) {
  return new Intl.DateTimeFormat('en-US', {
    timeZone: TZ,
    hour: 'numeric',
    minute: '2-digit',
    hour12: true,
  }).format(new Date(ms))
}

export function formatDayLabel(ms: number) {
  return new Intl.DateTimeFormat('en-US', {
    timeZone: TZ,
    weekday: 'short',
    month: 'short',
    day: 'numeric',
  }).format(new Date(ms))
}

/** Kalshi-style window: September 18, 11:30 – 11:45 AM CDT */
export function formatWindowRange(openAt?: number, closeAt?: number) {
  if (!openAt || !closeAt || !Number.isFinite(openAt) || !Number.isFinite(closeAt)) return '—'
  const open = new Intl.DateTimeFormat('en-US', {
    timeZone: TZ,
    month: 'long',
    day: 'numeric',
    hour: 'numeric',
    minute: '2-digit',
    hour12: true,
  }).format(new Date(openAt))
  const close = new Intl.DateTimeFormat('en-US', {
    timeZone: TZ,
    hour: 'numeric',
    minute: '2-digit',
    hour12: true,
  }).format(new Date(closeAt))
  return `${open} – ${close} CDT`
}

export function formatChartTick(ms: number, windowMs = 15 * 60_000) {
  return new Intl.DateTimeFormat('en-US', {
    timeZone: TZ,
    hour: 'numeric',
    minute: '2-digit',
    second: windowMs <= 5 * 60_000 ? '2-digit' : undefined,
    hour12: true,
  }).format(new Date(ms))
}

export function startOfChicagoDay(ms: number) {
  const p = parts(ms)
  const guess = Date.UTC(p.year, p.month - 1, p.day, 6, 0, 0)
  const off = chicagoOffsetMs(guess)
  return Date.UTC(p.year, p.month - 1, p.day, 0, 0, 0) - off
}

export function align15(ms: number) {
  const p = parts(ms)
  const minute = Math.floor(p.minute / 15) * 15
  const guess = Date.UTC(p.year, p.month - 1, p.day, p.hour, minute, 0)
  const off = chicagoOffsetMs(guess)
  return Date.UTC(p.year, p.month - 1, p.day, p.hour, minute, 0) - off
}

export function addChicagoDays(ms: number, days: number) {
  return startOfChicagoDay(ms + days * 86_400_000 + 12 * 3_600_000)
}

export function slotsForDay(dayMs: number) {
  const start = startOfChicagoDay(dayMs)
  const out: number[] = []
  for (let i = 0; i < 96; i++) out.push(start + i * 15 * 60_000)
  return out
}

export function upcomingDays(now = Date.now(), count = 7) {
  const today = startOfChicagoDay(now)
  return Array.from({ length: count }, (_, i) => {
    const t = addChicagoDays(today, i)
    return { t, key: dayKey(t), label: formatDayLabel(t), weekday: weekdayName(t) }
  })
}
